src/couch_views/src/couch_views_batch_impl.erl

```diff
@@ -0,0 +1,147 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+% http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_views_batch_impl).
+
+-behavior(couch_views_batch).
+
+
+-export([
+    start/2,
+    success/4,
+    failure/2
+]).
+
+
+-include("couch_mrview/include/couch_mrview.hrl").
+
+
+-record(batch_st, {
+    start_time,
+    size,
+    state = search,
+    search_incr,
+    sense_incr,
+    max_tx_size,
+    max_tx_time_msec,
+    threshold_penalty
+}).
+
+
+-spec start(
+        Mrst::#mrst{},
+        State::term()
+    ) -> {NewState::term(), BatchSize::pos_integer()}.
+start(Mrst, undefined) ->
+    St = #batch_st{
+        size = get_config("batch_initial_size", "100"),
+        search_incr = get_config("batch_search_increment", "500"),
+        sense_incr = get_config("batch_sense_increment", "100"),
+        max_tx_size = get_config("batch_max_tx_size", "9000000"),
+        max_tx_time_msec = get_config("batch_max_tx_time_msec", "4500"),
+        threshold_penalty = get_config("batch_threshold_penalty", "0.2")
```

Comment:
Minor nit: maybe set `state = search` explicitly here (as well or just here), just so all the state initialization is in one place and it's obvious that it always starts in `search`, stays there until it hits the threshold, then remains in `sense` after that.

Reply:
Done.